### ISSUE TITLE
Support parquet as a loader format

### ISSUE THREAD

**Issue by kszlim** [vega/vega#3961]
Request:
Support loading parquet as a source datatype in vega-loader

Envisioned solution:

```javascript
{
  "data": [
    {
      "name": "my_data",
      "format": {"type": "parquet"},
      "url": "https://path/to/parquet.parquet",
    }
  ]
}

```

This could be accomplished via https://github.com/kylebarron/parquet-wasm.

I'm not sure, but would there be a way to push down the required columns into the data loader, ie. whenever a user in the frontend wants to change what they're visualizing, could you get those columns pushed down (so that you don't have to load/read/decode the entire parquet file?).

This would make vega super bleeding edge/powerful.

**Comment by domoritz** [comment]
Vega fusion already does something like this for Vega in python. Nonetheless, it would be cool to adopt this for js as well.

**Comment by domoritz** [comment]
You could develop a loader similar to the Vega arrow loader as an external package. To you want to take a stab?

**Comment by domoritz** [comment]
Start small. You could manually project columns in the loader as a start.

**Comment by dangotbanned** [comment]
https://github.com/hyparam/hyparquet seems to be a lightweight alternative to https://github.com/kylebarron/parquet-wasm

> **Dependency-free:** Hyparquet has zero dependencies, making it lightweight and easy to use in any JavaScript project. Only 9.2KB min.gz!

IIUC, they'd be comparing **9.2 KB** vs **456 KB**.
You'd also somehow get [snappy](https://github.com/hyparam/hyparquet?tab=readme-ov-file#compression) compression support, despite the tiny size.

Seems like that might be a benefit over [uwdata/flechette](https://github.com/uwdata/flechette)'s `.arrow` support - which only works on uncompressed data.
- https://github.com/vega/vega-datasets/issues/655

**Comment by domoritz** [comment]
That's super cool. I'm hesitant to rely on the package in core for now but could be worth making an extension. 

The compressor package is more than 100kb, though: https://github.com/hyparam/hyparquet-compressors.

**Comment by dangotbanned** [comment]
> That's super cool. I'm hesitant to rely on the package in core for now but could be worth making an extension.
> 
> The compressor package is more than 100kb, though: [hyparam/hyparquet-compressors](https://github.com/hyparam/hyparquet-compressors?rgh-link-date=2025-02-11T12%3A56%3A43.000Z).

Hey @domoritz, yeah I'm not too sure on what these package sizes mean in JS-land 😅.

<details><summary>hyparquet vs hyparquet-compressors</summary>
<p>

![Image](https://github.com/user-attachments/assets/a8412938-57fd-4e55-97d0-ce70a1814e9e)

</p>
</details> 



The standalone (*snappy*) version would still give a huge reduction in file size vs uncompressed `.arrow`

https://github.com/vega/vega-datasets/issues/627#issuecomment-2474228235

Running this again, w/ the changes since (https://github.com/vega/vega-datasets/issues/655) I got:

| file_name                  | size (kb) |
| -------------------------- | --------- |
| flights-3m.arrow           | 128_908   |
| flights-3m_snappy.parquet  | 15_577    |
| flights-3m_zstd-22.parquet | 13_177    |

**Comment by dangotbanned** [comment]
@domoritz, am I understanding correctly that `.parquet` support *might* look like [vega-loader-arrow/src/index.js](https://github.com/vega/vega-loader-arrow/blob/1ff3cb919703265e569bc8892010da82dc553181/src/index.js) but wrapping another library?

I was expecting this to be a much more involved project

```js
import { tableFromIPC } from '@uwdata/flechette';

/**
 * Load a data set in Apache Arrow format for use in Vega.
 * @param {*} data Either an already parsed table (created by the
 *  `@uwdata/flechette` or `apache-arrow` libraries) or Arrow IPC
 *  binary data as an `ArrayBuffer`, `Uint8Array`, or `Uint8Array[]`.
 * @returns {Record<string,any>[]} An array of data objects representing
 *  rows of a data table.
 */
export default function arrow(data) {
  return (isArrowTable(data) ? data : decodeIPC(data)).toArray();
}

function isArrowTable(data) {
  return data && data.schema && Array.isArray(data.schema.fields)
    && typeof data.toArray === 'function';
}

function decodeIPC(data, options = { useProxy: true }) {
  return tableFromIPC(data, options);
}

arrow.responseType = 'arrayBuffer';
```

**Comment by domoritz** [comment]
Yep. Should be easy. If you want to  help set that up, that would be great. 

I was thinking of maybe bundling flechette with Vega in a coming version but haven't decided yet.

**Comment by jheer** [comment]
@dangotbanned, yes the Apache Arrow loader is pretty simple, as Flechette provides the needed (and efficient) routines for extracting data into the "array of objects" format that Vega expects. Hopefully something similarly easy is achievable with one of the existing Parquet libraries.

I believe the hyparquet library also supports efficient reading over the network using HTTP range requests, so it would be nice if Vega could also take advantage of that.

BTW, more generally I think adding a Parquet loader as a pluggable extension is a great idea. It is a better option for persistent file storage than Arrow (which instead particularly shines for interprocess communication).

**Comment by dangotbanned** [comment]
> Yep. Should be easy. If you want to help set that up, that would be great.

@domoritz well I'm intrigued, can't promise I'd be able to pull it off, but interested.
It seems to support [returning row-oriented data](https://github.com/hyparam/hyparquet?tab=readme-ov-file#returned-row-format)

>  believe the hyparquet library also supports efficient reading over the network using HTTP range requests, so it would be nice if Vega could also take advantage of that.

@jheer yeah I think that's mentioned in https://github.com/hyparam/hyparquet?tab=readme-ov-file#asyncbuffer

---

While I was looking around the repo, I found [this discussion](https://github.com/hyparam/hyparquet/issues/6) between the authors of [hyparquet](https://github.com/hyparam/hyparquet?rgh-link-date=2025-02-11T10%3A45%3A35.000Z) and [parquet-wasm](https://github.com/kylebarron/parquet-wasm?rgh-link-date=2025-02-11T10%3A45%3A35.000Z). 
Found it an interesting read - comparing the benefits of both

**Comment by domoritz** [comment]
I started an experimental implementation in https://github.com/vega/vega-loader-parquet. The main issue with async code so it doesn't work yet.
